### PR TITLE
Change From Clause indentations to fix L003 rule on template blocks

### PR DESCRIPTION
**Pull request by zidder**
### Brief summary of the change made

makes progess on #3496

Indent after `FROM` Key word, not before `FromExpression`.
This makes `autofix` examples nicer, and in case of `indented_joins=true`, it's consistent (and in my opinion, ideal). 
Two yaml tests show, that template from expression must be indented, and group by clause after that should not (didn't work on main branch).

### Are there any other side effects of this change that we should be aware of?
The `FromExpression` has Dedent before Indent, the corresponding Indent and Dedent are in the `FromClause`. In case there are several `FromExpression`s, the indent of the first `FromExpression` is dedented in the second `FromExpression`.

### Pull Request checklist
- [X] Please confirm you have completed any of the necessary steps below.

- Included test cases to demonstrate any code changes, which may be one or more of the following:
  - `.yml` rule test cases in `test/fixtures/rules/std_rule_cases`.
  - `.sql`/`.yml` parser test cases in `test/fixtures/dialects` (note YML files can be auto generated with `tox -e generate-fixture-yml`).
  - Full autofix test cases in `test/fixtures/linter/autofix`.
  - Other.
- Added appropriate documentation for the change.
- Created GitHub issues for any relevant followup/future enhancements if appropriate.
  - The issue is mentioned in #3496 , and several (#2677, #3133) can serve as a good followups. Can be solved in case we can dedent after empty template segment, or take this into consideration when linting those empty sections.

### CODE REVIEW

**Comment by zidder** [comment]
Can you help me fix the `mypy` issue in `dialect_mysql.py`? :)

**Comment by zidder** [comment]
Also, this introduces possibility of subsequent (indent, dedent) or (dedent, indent) segments. I think it would be better to remove those, even if they're in different segments (but are subsequent in raw segments).

**Comment by barrywhart** [comment]
This looks good to me. Tagging @tunetheweb for review, as he is more familiar with reviewing dialect changes.

File: src/sqlfluff/dialects/dialect_mysql.py [hunk 1]
```diff
@@ -531,6 +534,18 @@ class InsertStatementSegment(BaseSegment):
...
+
+    type = "indented_from_expression"
+
+    match_grammar = Sequence(
+        Indent,
+        Ref("FromExpressionSegment"),
+        Dedent,
+    )
+
+
```
Comment (tunetheweb) [inline src/sqlfluff/dialects/dialect_mysql.py:548]: I don't see the need for this class, and the extra level it creates in the parse tree. I agree it is used twice, but think cleaner to just `Indent`/`Dedent` directly in both those occasions - like the other dialects do.

This would also reduce some of the changes in these PR (including most of the mysql YAML changes?).

```suggestion
```

File: src/sqlfluff/dialects/dialect_mysql.py [hunk 2]
```diff
@@ -543,7 +558,7 @@ class DeleteUsingClauseSegment(BaseSegment):
     parse_grammar = Sequence(
         "USING",
         Delimited(
-            Ref("FromExpressionSegment"),
+            Ref("IndentedFromExpressionSegment"),
```
Comment (tunetheweb) [inline src/sqlfluff/dialects/dialect_mysql.py:561]:
```suggestion
            Indent,
            Ref("FromExpressionSegment"),
            Dedent,
```

File: src/sqlfluff/dialects/dialect_mysql.py [hunk 3]
```diff
@@ -2008,7 +2023,7 @@ class UpdateStatementSegment(BaseSegment):
         "UPDATE",
         Ref.keyword("LOW_PRIORITY", optional=True),
         Ref.keyword("IGNORE", optional=True),
-        Delimited(Ref("TableReferenceSegment"), Ref("FromExpressionSegment")),
+        Delimited(Ref("TableReferenceSegment"), Ref("IndentedFromExpressionSegment")),
```
Comment (tunetheweb) [inline src/sqlfluff/dialects/dialect_mysql.py:2026]:
```suggestion
        Delimited(Ref("TableReferenceSegment"), Indent, Ref("FromExpressionSegment"), Dedent),
```

File: src/sqlfluff/rules/L003.py [hunk 4]
```diff
@@ -79,10 +79,16 @@ def finalise(self, line_no: int, templated_file: Optional[TemplatedFile]):
...
+            copied_line_buffer, templated_file
+        )
         has_placeholder = any(
             elem.is_type("placeholder") for elem in copied_line_buffer
         )
-        is_empty_line = not has_code_segment and not has_placeholder
+
+        is_empty_line = not has_code_segment and (
+            not has_placeholder or templated_line_type == "end"
+        )
```
Comment (tunetheweb) [inline src/sqlfluff/rules/L003.py:91]: Why is this change needed? Should we have a comment here to explain?

File: test/fixtures/linter/autofix/bigquery/003_templating/after.sql [hunk 5]
```diff
@@ -14,15 +14,15 @@ SELECT
...
-        -- This next section gets very strange with forced template_blocks_indent
-        {{action}}_raw_effect_sizes
-    {% else %}
-    JOIN
+    {% for action in considered_actions %}
+        {% if loop.first %}
+            -- This next section gets very strange with forced template_blocks_indent
+            {{action}}_raw_effect_sizes
+        {% else %}
+        JOIN
```
Comment (tunetheweb) [inline test/fixtures/linter/autofix/bigquery/003_templating/after.sql:22]: Is this not indented too much now?

File: test/fixtures/rules/std_rule_cases/L003.yml [hunk 6]
```diff
@@ -1100,3 +1100,51 @@ test_tsql_bubble_up_newline_after_fix:
...
+            1 AS col
+        {% else %}
+            2 AS col
+        {% endif %}
+    FROM
+        {% if true %}
+            a
+        {% else %}
+        b -- won't be fixed as it's in else block
+    {% endif %} -- won't indent, as it is outside from block
```
Comment (tunetheweb) [inline test/fixtures/rules/std_rule_cases/L003.yml:1133]: Not sure I like this. While I agree it's difficult to make the `else` clause indent, surely the `{% endif %}` should be recognised as part of the `FROM` block?

File: src/sqlfluff/dialects/dialect_ansi.py [hunk 7]
```diff
@@ -1280,6 +1279,7 @@ class FromExpressionSegment(BaseSegment):
             Ref("JoinClauseSegment"), Ref("JoinLikeClauseGrammar"), optional=True
         ),
         Conditional(Dedent, indented_joins=True),
+        Indent,
```
Comment (tunetheweb) [inline src/sqlfluff/dialects/dialect_ansi.py:1282]: Not sure about this change - see comment later.

File: test/fixtures/templater/jinja_l_metas/001.yml [hunk 8]
```diff
@@ -28,16 +28,18 @@ file:
...
               - from_expression:
-                  indent: ''
                   from_expression_element:
                     table_expression:
                       table_reference:
                       - identifier: DATAHUB
                       - dot: .
                       - identifier: SNAPSHOT_DAILY
                   dedent: ''
+                  indent: ''
```
Comment (tunetheweb) [inline test/fixtures/templater/jinja_l_metas/001.yml:41]: I don't understand why we have these two? A `dedent` immediately followed by an `indent`? Is this because of the first change above on 1271/1282?

**Review by tunetheweb — CHANGES_REQUESTED** [review]
Think we should revert the mysql changes, and have questions on the others.

**Comment by zidder** [comment]
Sorry, can't find the right way to implement this :(
